### PR TITLE
ARROW-3191: [Java] Make ArrowBuf work with arbitrary underlying memory

### PR DESCRIPTION
**Pull request by siddharthteotia**
This patch has the following goals:

(1) Make ArrowBuf work with any arbitrary memory. 
(2) Decouple the usage of data get/set in ArrowBuf and memory accounting, reference management, ownership etc.

Changes

(1) A ReferenceManager interface that can be provided to ArrowBuf. This allows the users to provide their own custom implementation of reference management or it can be a NO-OP.
(2) All the accounting, ownership, reference related APIs have been moved to the default implementation of ReferenceManager -- BufferLedger, AllocationManager
(3) ArrowBuf is now literally an abstraction over some user provided underlying memory chunk. All it needs is starting virtual address and length of data to access along with a user provided implementation of ReferenceManager.
(4) ArrowBuf no longer extends or implements any of Netty's buffer interfaces. Thus all of the extra and unused APIs have been removed and it just provides simple get/set.

There is quite a bit of cleanup that needs to be done since some APIs have been moved out of ArrowBuf. So the caller code needs to change. They are likely going to be boilerplate changes but I would like to do them once we have consensus on the major set of changes here and the decoupling between ArrowBuf usage and reference management.

So the code doesn't compile yet because of the above mentioned reason. Secondly, there are a few things that I have removed assuming they are not being used  -- like BufferManager in ArrowBuf. I am still evaluating its usage. So there a few TODOs in code for these reasons.

Raising PR before the code is complete to get feedback on the important set of changes.

### CODE REVIEW

**Comment by pearu** [comment]
> Make ArrowBuf work with any arbitrary memory.

What does "arbitrary memory" mean? Would this include also device memory?

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 1]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
+   *
+   * So now to get/set data, we will do => addr + index
+   * This logic is put in method addr(index) and is frequently
+   * used in get/set data methods to compute the absolute
+   * byte address for get/set operation in the underlying chunk
+   *
+   * @param index the index at which we the user wants to read/write
+   * @return the absolute address within the memro
+   */
+  private long addr(int index) {
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: what do you think about changing the API to take LONGs here instead?  Or do you think this should be a separate PR?

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 2]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
-    final long addr = addr(index);
-    return (PlatformDependent.getByte(addr) & 0xff) |
-        (Short.reverseBytes(PlatformDependent.getShort(addr + 1)) & 0xffff) << 8;
+  public char getChar(int index) {
+    return (char) getShort(index);
   }
 
-  @Override
-  public ArrowBuf setShort(int index, int value) {
+  public ArrowBuf setChar(int index, int value) {
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: can we change this to either take a char, or replicate documentation saying which bits are ignored.

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 3]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
-  }
-
-  @Override
-  public ArrowBuf getBytes(int index, ByteBuffer dst) {
-    udle.getBytes(index + offset, dst);
-    return this;
-  }
-
-  @Override
   public ArrowBuf setByte(int index, int value) {
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: eliminate or document.

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 4]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
     chk(index, 1);
     PlatformDependent.putByte(addr(index), (byte) value);
     return this;
   }
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: Please try to add javadoc to each method

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 5]
```diff
@@ -403,89 +190,28 @@ public ArrowBuf slice(int index, int length) {
...
-    return slice(0, length);
-  }
-
-  @Override
-  public int nioBufferCount() {
-    return 1;
-  }
-
-  @Override
   public ByteBuffer nioBuffer() {
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: consider returning a collection here (if it isn't too big a change, even even we want to change to support 64 bit size memory in a separate PR).

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 2 above)
Comment (emkornfield) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: Actually, is there a use-case to store 2-byte characters in Arrow (I'm not thinking of one off the top of my head).

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 6]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
+  public ArrowBuf setLong(int index, long value) {
     chk(index, 8);
-    final long v = PlatformDependent.getLong(addr(index));
-    return v;
+    PlatformDependent.putLong(addr(index), value);
+    return this;
   }
 
-  @Override
   public float getFloat(int index) {
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: generally same question about changing the API to use Longs instead of Ints for all getters/settings

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 7]
```diff
@@ -17,144 +17,84 @@
...
  *
- * <p>Specifically, it serves as a facade over
- * {@linkplain UnsafeDirectLittleEndian} memory objects that hides the details
- * of raw memory addresses.
+ * (1) address - starting virtual address in the underlying memory
+ * chunk that this ArrowBuf has access to
  *
- * <p>ArrowBuf supports reference counting and ledgering to closely track where
- * memory is being used.
+ * (2) length - length (in bytes) in the underlying memory chunk
```
Comment (emkornfield) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: should this be tied to memory?  What if we want a file backed implementation?

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 1 above)
Comment (jacques-n) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: I don't think we should encumber this patch with that change.

Let's also have a discussion about that change on the mailing list before we make it. What use case is there where this is important?

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 7 above)
Comment (jacques-n) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: It definitely should be tied to a memory address. Whether that is backed by a memory mapped file will be up to the consumer with this change.

**Comment by jacques-n** [comment]
> > Make ArrowBuf work with any arbitrary memory.
> 
> What does "arbitrary memory" mean? Would this include also device memory?

It just means any memory address, not just those attached to a Netty allocator. Right now, ArrowBuf must be allocated by the Netty allocator which means that if a user allocated memory some other way, he is unable to connect it to Arrow. For example, the Java Plasma client has that problem right now.

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 4 above)
Comment (jacques-n) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: Let's make sure people are happy with overall shape/approach of patch before we spend time on this type of changes.

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 1 above)
Comment (emkornfield) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: I can bring this up for discussion on the mailing list but the suggestion is to bring this in line with the C++ side of things so large message batches can be read in java if desired see (https://issues.apache.org/jira/browse/ARROW-679)

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+   * memory is being associated with the target allocator or in other words allocation manager currently
+   * doesn't hold a mapping for the target allocator) or the current value of the reference count for
+   * the target allocator-reference manager combination + 1 in the case that the provided allocator
+   * already had an association to this underlying memory.
+   *
+   * @param srcBuffer source ArrowBuf
+   * @param targetAllocator The target allocator to create an association with.
+   * @return A new ArrowBuf which shares the same underlying memory as this ArrowBuf.
+   */
+  ArrowBuf retain (ArrowBuf srcBuffer, BufferAllocator targetAllocator);
```
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: this name doesn't seem to match up well with the documentation if an actual arrowbuf is being allocated.

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 9]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
+    long address = addr(index);
+    // word by word memcpy
+    while (toRead >= 8) {
+      dst.putLong(PlatformDependent.getLong(address));
+      address += 8;
+      toRead -= 8;
+    }
+    while (toRead > 0) {
+      dst.put(PlatformDependent.getByte(address));
+      --toRead;
```
Comment (BryanCutler) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: do you need a `++address` here?

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java [hunk 10]
```diff
@@ -499,461 +225,212 @@ public boolean equals(Object obj) {
...
+        PlatformDependent.copyMemory(src.array(), src.arrayOffset() + src.position(), dstAddress, (long)length);
+        src.position(src.position() + length);
+      } else {
+        while (length >= 8) {
+          PlatformDependent.putLong(dstAddress, src.getLong());
+          length -= 8;
+          dstAddress += 8;
+        }
+        while (length > 0) {
+          PlatformDependent.putByte(dstAddress, src.get());
```
Comment (BryanCutler) [inline java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: Do you need to decrement `length` and increment `dstAddress`?

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+   * Return the reference count
+   * @return reference count
+   */
+  int getRefCount();
+
+  /**
+   * Decrement this reference manager's reference count for the associated underlying
+   * memory. If the reference count drops to 0, it implies that ArrowBufs managed by this
+   * reference manager no longer need access to the underlying memory
+   * @return the new reference count
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: it returns a boolean though?

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  boolean release();
+
+  /**
+   * Decrement this reference manager's reference count for the associated underlying
+   * memory. If the reference count drops to 0, it implies that ArrowBufs managed by this
+   * reference manager no longer need access to the underlying memory
+   * @param increment the count to increase the reference count by
+   * @return the new reference count
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: same here

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+public interface ReferenceManager {
+
+  /**
+   * Return the reference count
+   * @return reference count
+   */
+  int getRefCount();
+
+  /**
+   * Decrement this reference manager's reference count for the associated underlying
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: Maybe say decrement by 1?

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+    return bufRefCnt.get();
+  }
+
+  /**
+   * Increment the ledger's reference count for the associated
+   * underlying memory chunk. All ArrowBufs managed by this ledger
+   * will share the ref count.
+   */
+  void increment() {
+    bufRefCnt.incrementAndGet();
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: could it be helpful to return the new count?

Comment (BryanCutler) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: I guess why do you need this since `retain()` is below?

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+    retain(1);
+  }
+
+  /**
+   * Increment the ledger's reference count for associated
+   * underlying memory chunk by the given amount
+   * @param increment amount to increase the reference count by
+   */
+  @Override
+  public void retain(int increment) {
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: Again here, could it be useful to return the new reference count?

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+    targetArrowBuf.writerIndex(srcArrowBuf.writerIndex());
+    return targetArrowBuf;
+  }
+
+  /**
+   * Transfer any balance the current ledger has to the target ledger. In the case
+   * that the current ledger holds no memory, no transfer is made to the new ledger.
+   *
+   * @param targetReferenceManager The ledger to transfer ownership account to.
+   * @return Whether transfer fit within target ledgers limits.
```
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: I guess if it's overlimit then it reallocates? Maybe you could clarify

**Review by BryanCutler — COMMENTED** [review]
+1 on the changes, looks great @siddharthteotia ! I just had some minor comments

**Comment by siddharthteotia** [comment]
Thanks @jacques-n , @BryanCutler, @emkornfield , @pearu for the comments so far. I will be addressing them soon. In my latest commit, I have made all the necessary changes in java code to work with new ArrowBuf and ReferenceManager interface. More importantly, there is a wrapper buffer NettyArrowBuf interface to comply with usage in RPC and Netty related code. 

It will be good to get feedback on that. As of now, the java modules build fine but I have to fix test failures. That is in progress.

**Comment by yuruiz** [comment]
please update the org/apache/arrow/memory/README.md to explain how does the ReferenceManager fit into Arrow Memory Management.

**Comment by siddharthteotia** [comment]
I got a clean build with my previous commit. The latest commit just has come cleanup and I expect it to be clean. Please review the changes. We should plan to merge this soon.

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 14 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: These interfaces are borrowed from how they were earlier in ArrowBuf. I didn't want to change or remove this reference count related APIs

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 15 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: May be. I just kept the way it was in previous implementation of ArrowBuf.

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 11 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: fixed.

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 12 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: fixed.

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 13 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: done

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 9 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: should be fixed in the latest commit

File: java/memory/src/main/java/io/netty/buffer/ArrowBuf.java (hunk 10 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/io/netty/buffer/ArrowBuf.java]: should be fixed in latest commit

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 13 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: done

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 12 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: done

File: java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java (hunk 11 above)
Comment (siddharthteotia) [reply java/memory/src/main/java/org/apache/arrow/memory/ReferenceManager.java]: yes the interface in old ArrowBuf had the semantics that it will return true if ref count has dropped to 0, false otherwise.

File: java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java (hunk 14 above)
Comment (BryanCutler) [reply java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java]: Ok, then I agree not to change it. I thought it was a new api.

**Comment by siddharthteotia** [comment]
Can this be merged?

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
merged to master, thanks @siddharthteotia

**Comment by siddharthteotia** [comment]
> merged to master, thanks @siddharthteotia

Thank you, @BryanCutler

**Comment by jacques-n** [comment]
I think @pravindra  and I both wanted to do another pass over this before merging. I guess we'll open up follow-on JIRAs to address additional comments.

**Review by jacques-n — COMMENTED** [review]
Some initial comments. Still going through it.

**Comment by BryanCutler** [comment]
Apologies @jacques-n and @pravindra if I merged too soon , I thought there were no more comments

**Comment by pravindra** [comment]
> Apologies @jacques-n and @pravindra if I merged too soon , I thought there were no more comments

no problem at all. we'll address issues, if any, in follow-up jiras.
